### PR TITLE
RMW QoS incoming buffer handling

### PR DESCRIPTION
**Pull request by pablogs9**
This PR implements the logic for handling the incoming RMW buffer depending on the QoS

TODO:
- [x] Req res tests
- [x] Backport to main and foxy

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ jobs:
         - name: Download dependencies
           run: |
             git clone -b foxy https://github.com/eProsima/Micro-CDR src/Micro-CDR
-            git clone -b foxy https://github.com/eProsima/Micro-XRCE-DDS-Client src/Micro-XRCE-DDS-Client
+            git clone -b feature/add_history_depth https://github.com/eProsima/Micro-XRCE-DDS-Client src/Micro-XRCE-DDS-Client
```
Comment (pablogs9) [inline .github/workflows/ci.yml:25]: When https://github.com/eProsima/Micro-XRCE-DDS-Client/pull/266 merged:
```suggestion
            git clone -b foxy https://github.com/eProsima/Micro-XRCE-DDS-Client src/Micro-XRCE-DDS-Client
```

**Comment by pablogs9** [comment]
@mergify backport main foxy

File: rmw_microxrcedds_c/test/test_reqres.cpp [hunk 2]
```diff
@@ -24,256 +24,497 @@
...
+      client, res_recv_data, sizeof(res_recv_data), taken,
+      response_header), RMW_RET_ERROR);
+
+  ASSERT_FALSE(taken);
+  ASSERT_NE(sequence_id, response_header.request_id.sequence_number);
+  ASSERT_NE(strcmp(res_data.c_str(), res_recv_data), 0);
 }
+
+// NOTE: Services are keep_last history 1 by default in the Agent
+// Those tests will require Client API modifications
```
Comment (Acuadros95) [inline rmw_microxrcedds_c/test/test_reqres.cpp:378]: Does this PR https://github.com/eProsima/Micro-XRCE-DDS-Client/pull/266, cover this functionality?

File: rmw_microxrcedds_c/src/callbacks.c [hunk 3]
```diff
@@ -70,16 +70,15 @@ void on_topic(
...
         RMW_SET_ERROR_MSG("Not available static buffer memory node");
         return;
       }
 
       rmw_uxrce_static_input_buffer_t * static_buffer =
         (rmw_uxrce_static_input_buffer_t *)memory_node->data;
-      static_buffer->owner = (void *) custom_subscription;
-      static_buffer->length = length;
 
       if (!ucdr_deserialize_array_uint8_t(
```
[the remark this answers is not in the document]
Comment (pablogs9) [reply rmw_microxrcedds_c/src/callbacks.c:83]: Fixed in 767b0fcccdd1e078732411f76fbab82f88836e71

[the remark this answers is not in the document]
Comment (pablogs9) [reply rmw_microxrcedds_c/src/callbacks.c:83]: Fixed in 767b0fcccdd1e078732411f76fbab82f88836e71

File: rmw_microxrcedds_c/test/test_reqres.cpp (hunk 2 above)
Comment (pablogs9) [reply rmw_microxrcedds_c/test/test_reqres.cpp:378]: Service has no QoS in XRCE binary mode. Pending.

File: rmw_microxrcedds_c/src/rmw_wait.c [hunk 4]
```diff
@@ -30,110 +31,111 @@ rmw_wait(
...
-          timeout = INT_MAX;
-          RMW_SET_ERROR_MSG("Wait timeout overflow");
-        }
-      }
-    }
-  } else {
+  uint64_t timeout = (uint64_t) rmw_time_total_nsec(*wait_timeout) / 1000000ULL;
+
+  if (rmw_time_equal(*wait_timeout, (rmw_time_t)RMW_DURATION_INFINITE)) {
     timeout = (uint64_t)UXR_TIMEOUT_INF;
```
[the remark this answers is not in the document]
Comment (pablogs9) [reply rmw_microxrcedds_c/src/rmw_wait.c]: Solved c94e6e76d5a9b1bbc4a94fe7aa8aaca1dcb76905

File: rmw_microxrcedds_c/src/callbacks.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Acuadros95) [reply rmw_microxrcedds_c/src/callbacks.c:83]: Add fix on request and reply callbacks: 8c914f0d5e38acbf5584e69006bec4aa7432e7b9

**Review by Acuadros95 — APPROVED** [review]

**Review by Acuadros95 — APPROVED** [review]

**Pull request merged**